### PR TITLE
Bring back data category tweaks

### LINKED ISSUE #751: Slowly add back "Data category tweaks"
I understand that this addon was too buggy to keep in the extension, but I personally found it very useful. I never experienced any problems with it, actually.

I'd suggest to take one of two courses of action:
1. Re-add the complete addon, but label it as Beta and/or add a warning about potential buggy behavior in the addon description. As bugs are reported and fixed, the Beta status could eventually be removed. (Or possibly not lol)
2. Re-add one component at a time. Whichever of the two (list category separation or local variable separation) was the least buggy could be added back first. If all goes well with that, add the second component on top of that. This could potentially help with finding bugs caused by both extensions being enabled at once.

### PR DESCRIPTION
**Pull request by GarboMuffin**
**Resolves**

Resolves #751 

**Changes**

image

This is a reimplementation of the data category tweaks addon that apparently caused various issues (#396). This PR uses a very different approach than the old version (#260) that is less invasive (no prototype hacks and doesn't reimplement the entire data category) so it might not have the same problems.

It supports both options that were supported by the old version (separate category for lists and separating global and local variables)

The old version changed the sorting logic for things like dropdowns and context menus when separating local variables is enabled, but this version does not. This is intentional - I don't want to touch prototypes or anything invasive until it's clear that what's already written doesn't cause issues.

This should be marked as beta until real users have had some time to try to break it because of all the issues with the old version.

**Tests**

This hasn't been thoroughly tested, but it seems to work fine for me. It needs more testing.

I did try out the old version and could consistently break it by making a new custom block, but this version is not affected by that particular issue, so it has at least 1 fewer bugs.

### CODE REVIEW

**Comment by scratchusernamemrtbts** [comment]
How about seperating cloud var and non-cloud var because some non-cloud var also have cloud stumble

**Comment by WorldLanguages** [comment]
TODO: if we keep the old addon ID, anyone that had this addon enabled when it was available will now have it enabled again. Is this intended?

File: addons/data-category-tweaks/userscript.js [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+  };
+
+  const separateLocalVariables = (workspace, toolboxXML) => {
+    const { variables, lists } = separateVariablesByType(toolboxXML);
+
+    const SMALL_GAP = 8;
+    const BIG_GAP = 24;
+
+    const makeLabel = (l10n) => {
+      const label = new DOMParser().parseFromString("<label></label>", "text/xml").firstChild;
```
Comment (TheColaber) [inline addons/data-category-tweaks/userscript.js:23]: Why do you need to do this to make a label?
Can's you just do
```js
const label = document.createElement("label")
```

File: addons/data-category-tweaks/userscript.js [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    const originalGetBlocksXML = vm.runtime.getBlocksXML;
+    vm.runtime.getBlocksXML = function (target) {
+      const result = originalGetBlocksXML.call(this, target);
+      hasSeparateListCategory = addon.settings.get("separateListCategory");
+      if (hasSeparateListCategory) {
+        result.push({
+          id: "data",
+          xml: `
+          <category
+            name="%{BKY_CATEGORY_VARIABLES}"
```
Comment (TheColaber) [inline addons/data-category-tweaks/userscript.js:143]: Whoops looks like this needs to be a "$"

Comment (GarboMuffin) [reply addons/data-category-tweaks/userscript.js:143]: % is correct. Blockly will parse and convert that to a translated string (https://github.com/LLK/scratch-gui/blob/ddd2fa06f2afa140a46ec03be91796ded861e65c/src/lib/make-toolbox-xml.js#L694-L704)

Comment (TheColaber) [reply addons/data-category-tweaks/userscript.js:143]: Ah... I see

**Review by GD-Distortion — APPROVED** [review]

**Comment by WorldLanguages** [comment]
I'll rename the addon ID so that no user gets a beta addon enabled by default after the update. We could keep the old addon ID and add some logic to the update routine but it's not worth it.

**Review by WorldLanguages — APPROVED** [review]

**Pull request merged**